File: backend/app/utils/account_type_groups.py
"""
Central registry of AccountType group memberships.

All other modules should import from here instead of defining their own
inline lists.  Groups are frozensets of AccountType enum members so they
can be used in ``in`` tests, SQLAlchemy ``.in_()``, and set operations.

DESIGN: Sets are composed from named atomic building blocks so that a
new type added to, say, EMPLOYER_PLAN_TYPES automatically propagates
to every set built on it.
"""

from app.models.account import AccountType, TaxTreatment

# ---------------------------------------------------------------------------
# Atomic building blocks
# ---------------------------------------------------------------------------

#: Pre-tax employer-sponsored plans (401k family)
EMPLOYER_PLAN_TYPES: frozenset[AccountType] = frozenset({
    AccountType.RETIREMENT_401K,
    AccountType.RETIREMENT_403B,
    AccountType.RETIREMENT_457B,
})

#: IRA family (Traditional, SEP, SIMPLE — all pre-tax by default)
TRADITIONAL_IRA_TYPES: frozenset[AccountType] = frozenset({
    AccountType.RETIREMENT_IRA,
    AccountType.RETIREMENT_SEP_IRA,
    AccountType.RETIREMENT_SIMPLE_IRA,
})

#: Roth family (after-tax, no RMD during owner's lifetime)
ROTH_TYPES: frozenset[AccountType] = frozenset({
    AccountType.RETIREMENT_ROTH,
})

#: Tax-free savings vehicles (not employer plans, not Roth)
TAX_FREE_SAVINGS_TYPES: frozenset[AccountType] = frozenset({
    AccountType.HSA,
    AccountType.RETIREMENT_529,
})

# ---------------------------------------------------------------------------
# Composed retirement groups
# ---------------------------------------------------------------------------

#: All retirement account types that hold investment assets
CORE_RETIREMENT_TYPES: frozenset[AccountType] = (
    EMPLOYER_PLAN_TYPES | TRADITIONAL_IRA_TYPES | ROTH_TYPES
)

#: All retirement types + tax-free savings
ALL_RETIREMENT_TYPES: frozenset[AccountType] = (
    CORE_RETIREMENT_TYPES | TAX_FREE_SAVINGS_TYPES
)

#: RMD-applicable accounts: pre-tax plans + PENSION (no ROTH, HSA, 529)
#: Per IRS rules: Roth IRAs do NOT require RMDs during owner's lifetime
RMD_ACCOUNT_TYPES: frozenset[AccountType] = (
    EMPLOYER_PLAN_TYPES | TRADITIONAL_IRA_TYPES | frozenset({AccountType.PENSION})
)

#: Types that show employer match UI
EMPLOYER_MATCH_TYPES: frozenset[AccountType] = EMPLOYER_PLAN_TYPES

#: Types eligible for Roth conversion analysis
#: Note: actual filter also checks tax_treatment != ROTH at query time
ROTH_CONVERSION_ELIGIBLE_TYPES: frozenset[AccountType] = CORE_RETIREMENT_TYPES

# ---------------------------------------------------------------------------
# Investment / holdings groups
# ---------------------------------------------------------------------------

#: Accounts that can hold securities/crypto.
#: MANUAL and OTHER included for user-tracked private investments.
INVESTMENT_ACCOUNT_TYPES: frozenset[AccountType] = (
    frozenset({
        AccountType.BROKERAGE,
        AccountType.CRYPTO,
        AccountType.MANUAL,
        AccountType.OTHER,
    })
    | ALL_RETIREMENT_TYPES
)

#: Cash account types (used in style-box Cash breakdown)
CASH_ACCOUNT_TYPES: frozenset[AccountType] = frozenset({
    AccountType.CHECKING,
    AccountType.SAVINGS,
    AccountType.MONEY_MARKET,
})

# ---------------------------------------------------------------------------
# Debt / liability groups
# ---------------------------------------------------------------------------

#: True liability types (balance is a debt, not an asset)
DEBT_ACCOUNT_TYPES: frozenset[AccountType] = frozenset({
    AccountType.CREDIT_CARD,
    AccountType.LOAN,
    AccountType.STUDENT_LOAN,
    AccountType.MORTGAGE,
})

#: Amortizing loan types (interest rate + scheduled payments).
#: Credit cards excluded — they use revolving logic, not amortization.
AMORTIZING_LOAN_TYPES: frozenset[AccountType] = frozenset({
    AccountType.MORTGAGE,
    AccountType.LOAN,
    AccountType.STUDENT_LOAN,
})

# ---------------------------------------------------------------------------
# Cash-flow exclusion groups
# ---------------------------------------------------------------------------

#: Manual accounts excluded from cash flow by default.
MANUAL_EXCLUDE_CASH_FLOW_TYPES: frozenset[AccountType] = (
    AMORTIZING_LOAN_TYPES | frozenset({AccountType.CREDIT_CARD})
)

#: Plaid-linked accounts excluded from cash flow by default.
#: Credit cards intentionally NOT excluded — purchases show in cash flow.
PLAID_EXCLUDE_CASH_FLOW_TYPES: frozenset[AccountType] = AMORTIZING_LOAN_TYPES

# ---------------------------------------------------------------------------
# Net worth exclusion
# ---------------------------------------------------------------------------

#: Types that default to EXCLUDED from net worth when include_in_networth is NULL.
#: Note: PRIVATE_EQUITY has a secondary rule (excluded unless public) handled in code.
NET_WORTH_EXCLUDED_BY_DEFAULT: frozenset[AccountType] = frozenset({
    AccountType.VEHICLE,
    AccountType.COLLECTIBLES,
    AccountType.OTHER,
    AccountType.MANUAL,
    AccountType.PENSION,
})

# ---------------------------------------------------------------------------
# Tax treatment defaults
# ---------------------------------------------------------------------------

#: Default TaxTreatment to assign when creating a new account with no explicit value.
TAX_TREATMENT_DEFAULTS: dict[AccountType, TaxTreatment] = {
    AccountType.RETIREMENT_401K:       TaxTreatment.PRE_TAX,
    AccountType.RETIREMENT_403B:       TaxTreatment.PRE_TAX,
    AccountType.RETIREMENT_457B:       TaxTreatment.PRE_TAX,
    AccountType.RETIREMENT_IRA:        TaxTreatment.PRE_TAX,
    AccountType.RETIREMENT_SEP_IRA:    TaxTreatment.PRE_TAX,
    AccountType.RETIREMENT_SIMPLE_IRA: TaxTreatment.PRE_TAX,
    AccountType.RETIREMENT_ROTH:       TaxTreatment.ROTH,
    AccountType.HSA:                   TaxTreatment.TAX_FREE,
    AccountType.RETIREMENT_529:        TaxTreatment.TAX_FREE,
}

File: frontend/src/constants/accountTypeGroups.ts
/**
 * Central registry of account type group memberships.
 *
 * All UI logic that tests which "bucket" an account_type falls into should
 * import from here. Changing a group here is sufficient — no need to update
 * every component that references the group.
 *
 * DESIGN: Arrays are composed from smaller named arrays (via spread) so
 * groups stay in sync. The `as const` suffix preserves literal types.
 */

import { AccountType } from '../types/account';

// ---------------------------------------------------------------------------
// Atomic building blocks (mirror the Python layer)
// ---------------------------------------------------------------------------

/** Pre-tax employer-sponsored plans (401k family) */
export const EMPLOYER_PLAN_TYPES = [
  AccountType.RETIREMENT_401K,
  AccountType.RETIREMENT_403B,
  AccountType.RETIREMENT_457B,
] as const;

/** IRA family (Traditional, SEP, SIMPLE — all pre-tax by default) */
export const TRADITIONAL_IRA_TYPES = [
  AccountType.RETIREMENT_IRA,
  AccountType.RETIREMENT_SEP_IRA,
  AccountType.RETIREMENT_SIMPLE_IRA,
] as const;

/** Roth family */
export const ROTH_TYPES = [
  AccountType.RETIREMENT_ROTH,
] as const;

/** Tax-free savings vehicles */
export const TAX_FREE_SAVINGS_TYPES = [
  AccountType.HSA,
  AccountType.RETIREMENT_529,
] as const;

// ---------------------------------------------------------------------------
// Composed groups
// ---------------------------------------------------------------------------

/** All retirement types + tax-free savings */
export const ALL_RETIREMENT_TYPES = [
  ...EMPLOYER_PLAN_TYPES,
  ...TRADITIONAL_IRA_TYPES,
  ...ROTH_TYPES,
  ...TAX_FREE_SAVINGS_TYPES,
] as const;

/** Manual accounts that make sense to schedule recurring contributions for. */
export const CONTRIBUTION_ACCOUNT_TYPES = [
  AccountType.SAVINGS,
  AccountType.BROKERAGE,
  ...EMPLOYER_PLAN_TYPES,
  ...TRADITIONAL_IRA_TYPES,
  ...ROTH_TYPES,
  ...TAX_FREE_SAVINGS_TYPES,
] as const;

/** Investment accounts that can have individual holdings tracked. */
export const HOLDINGS_ACCOUNT_TYPES = [
  AccountType.BROKERAGE,
  ...EMPLOYER_PLAN_TYPES,
  ...TRADITIONAL_IRA_TYPES,
  ...ROTH_TYPES,
  ...TAX_FREE_SAVINGS_TYPES,
  AccountType.CRYPTO,
] as const;

/** Asset accounts track a value, not a transaction flow. */
export const ASSET_ACCOUNT_TYPES = [
  AccountType.PROPERTY,
  AccountType.VEHICLE,
  AccountType.COLLECTIBLES,
  AccountType.PRECIOUS_METALS,
  AccountType.BUSINESS_EQUITY,
  AccountType.PRIVATE_EQUITY,
  AccountType.PRIVATE_DEBT,
  AccountType.BOND,
  AccountType.LIFE_INSURANCE_CASH_VALUE,
  AccountType.PENSION,
  AccountType.ANNUITY,
] as const;

/** Debt account types — balance is negative, shown with debt UI. */
export const DEBT_ACCOUNT_TYPES = [
  AccountType.CREDIT_CARD,
  AccountType.LOAN,
  AccountType.STUDENT_LOAN,
  AccountType.MORTGAGE,
] as const;

/** Types that show the tax treatment selector in account settings. */
export const TAX_TREATMENT_ACCOUNT_TYPES = [
  ...ALL_RETIREMENT_TYPES,
  AccountType.BROKERAGE,
] as const;

/** Types that show employer match UI. */
export const EMPLOYER_MATCH_TYPES = EMPLOYER_PLAN_TYPES;

// ---------------------------------------------------------------------------
// Sidebar grouping config
// ---------------------------------------------------------------------------

/** Maps every AccountType to a sidebar section label and sort order. */
export const ACCOUNT_TYPE_SIDEBAR_CONFIG: Record<string, { label: string; order: number }> = {
  [AccountType.CHECKING]:              { label: 'Cash', order: 1 },
  [AccountType.SAVINGS]:               { label: 'Cash', order: 1 },
  [AccountType.MONEY_MARKET]:          { label: 'Cash', order: 1 },
  [AccountType.CD]:                    { label: 'Cash', order: 1 },
  [AccountType.CREDIT_CARD]:           { label: 'Credit Cards', order: 2 },
  [AccountType.BROKERAGE]:             { label: 'Investments', order: 3 },
  [AccountType.PRIVATE_EQUITY]:        { label: 'Investments', order: 3 },
  [AccountType.CRYPTO]:                { label: 'Investments', order: 3 },
  [AccountType.RETIREMENT_401K]:       { label: 'Retirement', order: 4 },
  [AccountType.RETIREMENT_403B]:       { label: 'Retirement', order: 4 },
  [AccountType.RETIREMENT_457B]:       { label: 'Retirement', order: 4 },
  [AccountType.RETIREMENT_IRA]:        { label: 'Retirement', order: 4 },
  [AccountType.RETIREMENT_ROTH]:       { label: 'Retirement', order: 4 },
  [AccountType.RETIREMENT_SEP_IRA]:    { label: 'Retirement', order: 4 },
  [AccountType.RETIREMENT_SIMPLE_IRA]: { label: 'Retirement', order: 4 },
  [AccountType.RETIREMENT_529]:        { label: 'Retirement', order: 4 },
  [AccountType.HSA]:                   { label: 'Retirement', order: 4 },
  [AccountType.PENSION]:               { label: 'Retirement', order: 4 },
  [AccountType.LOAN]:                  { label: 'Loans', order: 5 },
  [AccountType.STUDENT_LOAN]:          { label: 'Loans', order: 5 },
  [AccountType.MORTGAGE]:              { label: 'Loans', order: 5 },
  [AccountType.PROPERTY]:              { label: 'Property', order: 6 },
  [AccountType.VEHICLE]:               { label: 'Property', order: 6 },
  [AccountType.BOND]:                  { label: 'Bonds & Securities', order: 7 },
  [AccountType.STOCK_OPTIONS]:         { label: 'Bonds & Securities', order: 7 },
  [AccountType.PRIVATE_DEBT]:          { label: 'Alternative', order: 8 },
  [AccountType.COLLECTIBLES]:          { label: 'Alternative', order: 8 },
  [AccountType.PRECIOUS_METALS]:       { label: 'Alternative', order: 8 },
  [AccountType.LIFE_INSURANCE_CASH_VALUE]: { label: 'Insurance', order: 9 },
  [AccountType.ANNUITY]:               { label: 'Insurance', order: 9 },
  [AccountType.BUSINESS_EQUITY]:       { label: 'Business', order: 10 },
  [AccountType.MANUAL]:                { label: 'Other', order: 11 },
  [AccountType.OTHER]:                 { label: 'Other', order: 11 },
};
